pilot/model/validation.go

```diff
@@ -1624,6 +1622,217 @@ func ValidateEndUserAuthenticationPolicySpecBinding(msg proto.Message) error {
 	return errs
 }
 
+// ValidateRouteRuleV2 checks that a v1alpha2 route rule is well-formed.
+func ValidateRouteRuleV2(msg proto.Message) (errs error) {
+	routeRule, ok := msg.(*routingv2.RouteRule)
+	if !ok {
+		return errors.New("cannot cast to v1alpha2 routing rule")
+	}
+
+	// TODO: routeRule.Gateways
+
+	if len(routeRule.Hosts) == 0 {
+		errs = multierror.Append(errs, errors.New("at least one host required"))
+	}
+
+	for _, http := range routeRule.Http {
+		// check for conflicts
+		if http.Redirect != nil {
+			if len(http.Route) > 0 {
+				errs = appendErrors(errs, errors.New("HTTP route cannot contain both route and redirect"))
+			}
+
+			if http.Fault != nil {
+				errs = appendErrors(errs, errors.New("HTTP route cannot contain both fault and redirect"))
+			}
+
+			if http.Rewrite != nil {
+				errs = appendErrors(errs, errors.New("HTTP route rule cannot contain both rewrite and redirect"))
+			}
+
+			if http.WebsocketUpgrade {
+				errs = appendErrors(errs, errors.New("WebSocket upgrade is not allowed on redirect rules")) // nolint: golint
+			}
+		}
+
+		for name := range http.AppendHeaders {
+			errs = appendErrors(errs, ValidateHTTPHeaderName(name))
+		}
+		errs = appendErrors(errs, validateCORSPolicy(http.CorsPolicy))
+		errs = appendErrors(errs, validateHTTPFaultInjection(http.Fault))
+
+		for _, match := range http.Match {
+			for name := range match.Headers {
+				errs = appendErrors(errs, ValidateHTTPHeaderName(name))
+			}
+
+			// TODO: validate once implemented
+			if match.Port != nil {
+				errs = appendErrors(errs, errors.New("HTTP match port has not been implemented"))
+			}
+
+			errs = appendErrors(errs, Labels(match.SourceLabels).Validate())
+		}
+		errs = appendErrors(errs, validateDestination(http.Mirror))
+		errs = appendErrors(errs, validateHTTPRedirect(http.Redirect))
+		errs = appendErrors(errs, validateHTTPRetry(http.Retries))
+		errs = appendErrors(errs, validateHTTPRewrite(http.Rewrite))
+		for _, route := range http.Route {
+			if route.Destination == nil {
+				errs = multierror.Append(errs, errors.New("destination is required"))
+			}
+			errs = appendErrors(errs, validateDestination(route.Destination))
+			errs = appendErrors(errs, ValidatePercent(route.Weight))
+		}
+		if http.Timeout != nil {
+			errs = appendErrors(errs, ValidateDuration(http.Timeout))
+		}
+	}
+
+	// TODO: validate once implemented
+	if len(routeRule.Tcp) > 0 {
+		errs = appendErrors(errs, errors.New("TCP route rules have not been implemented"))
+	}
+
+	return
```

Comment:
May be refactor into two functions like validateHTTPRoute, and validateTCPRoute.. and then call them from this function, per iteration in the for loop ?

Reply:
done